### ISSUE TITLE
More precise audit logs for actions taken by API Keys

### ISSUE THREAD

**Issue by tvernum** [elastic/elasticsearch#52551]
Related: https://github.com/elastic/elasticsearch/issues/52314

At the moment, if an action is authenticated via an API key, the audit logs do not tell the complete story of what happened:
1. The user's principal is the original principal, with no record of the API key id/name
2. We set the realm name `"_es_api_key"`, but dont return the original realm name so the original principal might be ambiguous (enchantress@dc and enchantress@marvel are different users, so enchantress@_es_api_key is abigiuous).
3. We audit role names, but they are not meaningful in an API key scenario (when we have specific privileges in use)
4. We don't record the authentication type, so you need to imply the type from the realm name

**Comment by tvernum** [comment]
I haven't marked this `team-discuss` because I think we should just sort it out, and most of it is straight forward.

The roles vs privileges issue is a tricky one that will need some thought, but adding new fields for API key info & authentication type isn't controversial.

**Comment by albertzaharovits** [comment]
After some digging, my plan for audit improvements for API Key authentication is to 

1) modify the `access_granted`, `access_denied` and `authentication_success` audit record types, as follows:
- add new `apikey.id` and `apikey.name` audit entry fields
- add a new `authentication.type` audit entry field with values such as token/realm/api_key
- remove the `user.roles` field as it currently contains a synthetic role name that is not informative of anything (except for the `authentication_success` event type, which does not contain this field)
- change the value of the `user.realm` field to reflect the realm of the user that created the API Key, instead of the current synthetic value of `_es_api_key` which is again not informative

I believe not maintaining values for the realm name and roles in the audit log _might_ break some downstream tooling, but these values should not be part of the output interface that we maintain because often they expose internals, which are flickery,  such as action and request type names, or, as in this case, hardcoded bogus names. They should've never been exposed in the first place.

2) make the authentication process generate an `authentication_failed` event when API Key authentication fails (it currently silently fails through to realm authentication). The fields would be the same as those from `authentication_success`, minus `apikey.name` and `user.realm` when they're not available because the API key does not exist.

Auditing of API Key Create, Update and Delete operations will be handled under a separate effort to audit Security changes in general.

**Comment by albertzaharovits** [comment]
Another possible change, that I don't plan on implementing but that I want to raise up for discussion, is whether we should change the `user.name` field to something like `owner.name`, or even remove the field altogether, when authn is performed via API Keys.

The owner provides the baseline of privileges for a given API Key, and, in a sense, API Keys are used to delegate access to the owner resources'.

Eventually, the `apikey.id` field could be used to reference the audit event of the API Key creation (which will be added at a latter time), providing all the informations about the API Key, including the owner user (in addition to the creator user, expiration and limiting privileges).

To summarise, I see 3 options for how to treat the owner of the API Key used for authn:

1. keep the existing `user.name` field (that's used for realm authn) to designate the owner of the API Key
2. remove the `user.name` field altogether, relying on the new `apikey.id` field to reference the (to-be-added) audit event for API Key creation (which includes the owner information)
3. remove the `user.name` field and introduce another field, `owner.name`.

I favour the first option because I _feel_ that API Key authentications are conceptually ascribed to the user that created the key. I don't have strong arguments, it's just a feeling based on the use cases where users create API Keys for services that must run under their context.

Does anyone from @elastic/es-security or @bytebilly have other thoughts on this matter?

**Comment by ywangd** [comment]
Technically, if API Key creation is audited comprehensively, a simple API key ID should suffice the tracibility requirement. But I understand there are some arguments in favor of having user information in later logs, especially it provides more information when you glance through the logs. If certain users do not understand how API key works, the information could be confusing but they are argubly "more". 

In terms of semantic accuracy, I do think `owner.name` is better. However, I am concerned that this would break existing log analysis toolings that users may already have. Also this could lead to discussions on what are qualified as `owner`. Should we rename it for token authentication as well? Even the serialised auth header could be considered as "owned" by an user. 

Therefore, I do think **option 1 is better** in practice. It could use help of improved documentation to clarify that its precise semantics depends on other fields (e.g. whether `apikey.id` exists). Improvement on documentation is needed regardlessly, e.g. `owner.name` also needs to be documented to clarify its exact meaning, so it is _not_ a downside.

**Comment by tvernum** [comment]
I agree with option 1, for similar reasons.

_If_ we were introducing entirely new audit events for API Key authentication, I would probably favour `owner.name` over `user.name`, but it would be a weak preference.

However, since we're not creating separate events (and I don't think we should), I think the consistency of a single field is preferable.
1. Existing tooling expects `user.name` to always be there. Changing that seems like unnecessary churn.
2. If an audit log processor is not aware of API Keys, then they will get some useful information from the `user.name` field. It's not perfect, but it's something meaningful on which to base an audit investigation.
3. If the entries are being ingested into a ES index, then I feel like a single field is preferable. It improves the lucene performance (no sparse fields) and also makes it possible to do useful aggregations.
4. For use cases like Kibana alerting where the API Key really is acting on behalf of the owner, the above points become more significant. There will be clusters whether the only use of API Keys are for alerting, and in those cases you can ignore the `apikey.id` for 99% of cases and just rely on `user.name` as a consistent pointer to _Who did this?_ without caring (at least initially) with the _How?_ question.

**Comment by jkakavas** [comment]
Agreed with the general approach and +1 on option 1 in your open question Albert. 

Even if we didn't do that, I think that since we are using both access tokens and api keys as bearer tokens, `owner` would also be semantically vague so I'd still favor `user`.

> The fields would be the same as those from authentication_success, minus apikey.name and user.realm when they're not available because the API key does not exist.

I think this is what you meant too, but just checking, the fields would be missing both when the apikey with that id doesn't exist and when the api key password was incorrect , right?

**Comment by tvernum** [comment]
@albertzaharovits When you have a chance, can you update this issue with a list of what's still outstanding? #58928 covered most of it, and I'm not sure if we can close this, or if we think there's more to do.

**Comment by albertzaharovits** [comment]
@tvernum I've double checked and I believe that the only thing still left for API key auditing are authentication failures.

Currently, if API Key authentication fails, for any reason, the authentication flow follows through the realm chain. If none of the realms succeed, which is what happens if the request only contains the API Key credentials header, the authentication fails as if no credentials are present. In this case an anonymous access denied audit event is emitted, eg:
```
{"type":"audit", "timestamp":"2020-09-29T12:08:43,228+0300", "node.id":"U2LDsAJVQfqCFZlVLe1rjQ", "event.type":"rest", "event.action":"anonymous_access_denied", "origin.type":"rest", "origin.address":"[::1]:57670", "url.path":"/index-b1/_doc/2", "url.query":"pretty", "request.method":"PUT", "request.id":"dgYxFOu6TQun9KIrm6QwDQ"}
```

This behaviour is similar to that of bearer tokens, but, comparatively, those are verified more stringently. For API Key authentication, format/decoding errors are interpreted as _authn unsuccessful_ errors.

My proposal is to audit such failures similarly to how we audit the realm authentication failures, but to also have these emitted by default (unlike realm authentication failures).

I'll add the `team-discuss` label to get a check on the approach.

**Comment by albertzaharovits** [comment]
We've discussed this and agreed that authentication failure events should be emitted when API key and tokens verification fails (before authn falls through to realms).
